parser/validate.go

```diff
@@ -456,6 +456,12 @@ func validateField(res *result, syntax syntaxType, name protoreflect.FullName, f
 	}
 
 	node := res.FieldNode(fld)
+	if fld.GetNumber() == 0 && node.FieldTag() == nil {
```

Comment:
Above, instead of setting the tag to zero in this case, you could leave the tag as nil (it's a pointer since it's a proto2 optional field). That way you can safely change the `FieldTag()` method to _not_ return nil and this check can still work -- it would instead check if `fld.Number == nil`.